internal/core/software_renderer.rs

```diff
@@ -1008,6 +1050,77 @@ impl<'a, T: ProcessScene> SceneBuilder<'a, T> {
         };
     }
 
+    fn draw_text_paragraph<'b, Font: AbstractFont>(
+        &mut self,
+        paragraph: TextParagraphLayout<'b, Font>,
+        physical_clip: euclid::Rect<f32, PhysicalPx>,
+        offset: euclid::Vector2D<f32, PhysicalPx>,
+        color: Color,
+    ) where
+        Font: crate::textlayout::TextShaper<Length = PhysicalLength>,
+        Font: GlyphRenderer,
+    {
+        paragraph.layout_lines(|glyphs, line_x, line_y| {
+            let baseline_y = line_y + paragraph.layout.font.ascent();
+            while let Some(positioned_glyph) = glyphs.next() {
+                let glyph = paragraph.layout.font.render_glyph(positioned_glyph.glyph_id);
```

Comment:
If I understand correctly, this will always re-allocate the data in a new Rc  So Rc is not even required. Just a Box<[u8]> would work.
It would be nice if the font data could be cached somehow.    Becasue that's one allocation per frame per glyphs, quite a lot.
Anyway, this is just for the emulator right now so i guess that's fine.

Reply:
In the bitmap font case not allocation happens, and in the case of the vector font most of the time the glyph comes as cloned Rc from the glyph cache.

If you try the initial commit where there's no cache yet you'll notice that it's suuuuuuuuuuuper slow. With the cache it gets a lot better. There's still some slowness left, but I haven't profiled that yet. Could be text layout related (font loading at shaping time) or something else. Hard to say if it's a regression since I had never seen the widgets gallery with the software renderer before :)